metaflow/plugins/kfp/kfp_split_contexts.py

```diff
@@ -0,0 +1,189 @@
+    seen_steps = {"start"}  # Set of seen steps
+    task_id = 0
+    while len(steps_queue) > 0:
+        current_step = steps_queue.pop(0)
+        node = graph.nodes[current_step]
+        task_id += 1
+        step_to_task_id[current_step] = task_id
+
+        for step in node.out_funcs:
+            if step not in seen_steps:
+                steps_queue.append(step)
+                seen_steps.add(step)
+
+    return step_to_task_id
+
+
+class KfpSplitContext(object):
+    """
+    passed_in_split_indexes is a string of foreach split_index ordinals.
+    A nested foreach appends the new split index ordinal with a "_" separator.
+    Example:
+        0_1 -> 0th index of outer foreach and 1th index of inner foreach
+        1_0 -> 1th index of outer foreach and 0th index of inner foreach
+
+    Please see metaflow_nested_foreach.ipynb for more.
+    """
+
+    def __init__(
+        self,
+        graph: FlowGraph,
+        step_name: str,
+        run_id: str,
+        datastore: MetaflowDataStore,
+        logger: Callable,
+    ):
+        self.graph = graph
+        self.step_name = step_name
+        self.run_id = run_id
+        self.logger = logger
+        self.node = graph[step_name]
+        self.flow_root = datastore.make_path(graph.name, run_id)
+
+    def build_context_dict(self, flow: FlowSpec) -> Dict[str, str]:
+        """
+        Returns a dict with
+        {
+            task_id: The task_id of this step
+            foreach_splits: <PASSED_IN_SPLIT_INDEXES>_index where index is ordinal of the split
+        }
+        """
+        if self.node.type == "foreach":
+            passed_in_split_indexes = os.environ[PASSED_IN_SPLIT_INDEXES_ENV_NAME]
+
+            # The splits are fed to kfp.ParallelFor to downstream steps as
+            # "passed_in_split_indexes" variable and become the step task_id
+            # Example: 0_3_1
+            foreach_splits = [
+                f"{passed_in_split_indexes}{SPLIT_INDEX_SEPARATOR}{split_index}".strip(
+                    SPLIT_INDEX_SEPARATOR
+                )
+                for split_index in range(0, flow._foreach_num_splits)
+            ]
+
+            return {
+                "task_id": current.task_id,
+                "foreach_splits": foreach_splits,
+            }
+        else:
+            return {
+                "task_id": current.task_id,
+            }
+
+    def get_parent_context(
+        self,
+        parent_context_step_name: str,
+        current_node: DAGNode,
+        passed_in_split_indexes: str,
+    ) -> Dict[str, str]:
+        """
+        Used by compute_input_paths() to access the parent context saved in step_kfp_output_path
+
+        Returns:
+            Task context dict built by build_context_dict() that was saved to S3 by kfp_decorator.
+        """
+        self.logger(parent_context_step_name, head="--")
+
+        context_node_task_id = str(
+            graph_to_task_ids(self.graph)[parent_context_step_name]
```

Comment:
Looks like this `graph_to_task_ids` will be recomputed on every call to `get_parent_context` is this something we can move to the constructor to not pay that compute cost?